File: engine/city.lua
local City = {}
City.__index = City

function City:new(x, y, tp, biome)
   local self = setmetatable({}, City)
   self.x, self.y = x, y
   local layout_number = math.random(1,#CityLayout[tp])
   local layout = CityLayout[tp][layout_number]
   self.w, self.h = layout.w, layout.h
   self.buildings = {}
   self.biome = biome
   self:__load_layout(layout)
   return self
end

function City:tiles(x, y)
   for _, bd in ipairs(self.buildings) do
      if x >= bd.x and x < (bd.x+bd.w) and y >= bd.y and y < (bd.y+bd.h) then
         return bd:tiles(x-bd.x, y-bd.y, self.biome)
      end
   end
   return { self.biome }
end

-------------
-- PRIVATE --
-------------

function City:__tostring()
   return '[City]'
end

function City:__load_layout(layout)
   for _,bd in ipairs(layout.buildings) do
      local building = Building:new(bd.layout, bd.x, bd.y)
      if building.x+building.w > self.w or building.y+building.h > self.h then
         error('Building outside city.')
      end
      self.buildings[#self.buildings+1] = building
   end
end

return City

-- vim: ts=3:sw=3:sts=3:expandtab

File: engine/city.cc
#include "engine/city.h"

#include <cassert>

#include "./main.h"
#include "engine/world.h"

City::City(int x, int y, CityType type, int n)
    : X(x), Y(y), layout(CityLayouts[CityPair(type, n)]), buildings()
{
    for(auto const& bpos : layout.buildings) {
        buildings.emplace_back(
                static_cast<int>(bpos.second.X()), 
                static_cast<int>(bpos.second.Y()), 
                bpos.first.first, bpos.first.second);
    }
}


int 
City::Tiles(Block (&block)[10], int x, int y) const
{
    for(auto const& b : buildings) {
        if(x >= b.X && x < b.X+b.W()
        && y >= b.Y && y < b.Y+b.H()) {
            return b.Tiles(block, x - b.X, y - b.Y);
        }
    }
    block[0] = Block::EMPTY;
    return 1;
}


CityType 
City::Type(Block b)
{
    if(b == Block::GRASS) {
        return CityType::AGRICULTURAL;
    } else if(b == Block::SNOW) {
        return CityType::REFINERY;
    } else if(b == Block::BARE) {
        return CityType::MINING;
    } else if(b == Block::TEMPERATE_FOREST || b == Block::TROPICAL_FOREST) {
        return CityType::FORESTAL;
    } else if(b == Block::DESERT || b == Block::BEACH) {
        return CityType::CHEMICAL;
    } else {
        return static_cast<CityType>(world->Random() * CityType::__COUNT__);
    }
}


const std::vector<WorkerPair> 
City::Workers() const
{
    std::vector<WorkerPair> pairs;

    for(auto const& building: buildings) {
        for(auto const& worker: building.Workers()) {
            pairs.push_back({ worker.first,
                    Point(worker.second.X() + building.X + X,
                          worker.second.Y() + building.Y + Y) });
        }
    }

    return pairs;
}


// vim: ts=4:sw=4:sts=4:expandtab
